server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -18,58 +18,74 @@
 
-
 /**
- * This component is responsible for connecting to nodes once they are added to the cluster state, and disconnect when they are
- * removed. Also, it periodically checks that all connections are still open and if needed restores them.
- * Note that this component is *not* responsible for removing nodes from the cluster if they disconnect / do not respond
- * to pings. This is done by {@link FollowersChecker}. Master fault detection is done by {@link LeaderChecker}.
+ * This component is responsible for maintaining connections from this node to all the nodes listed in the cluster state, and for
+ * disconnecting from nodes once they are removed from the cluster state. It periodically checks that all connections are still open and
+ * restores them if needed. Note that this component is *not* responsible for removing nodes from the cluster state if they disconnect or
+ * are unresponsive: this is the job of the master's fault detection components, particularly {@link FollowersChecker}.
+ * <p>
+ * The {@link NodeConnectionsService#connectToNodes(DiscoveryNodes, Runnable)} and {@link
+ * NodeConnectionsService#disconnectFromNodesExcept(DiscoveryNodes)} methods are called on the {@link ClusterApplier} thread. This component
+ * allows the {@code ClusterApplier} to block on forming connections to _new_ nodes, because the rest of the system treats a missing
+ * connection with another node in the cluster state as an exceptional condition and we don't want this to happen to new nodes. However we
+ * need not block on re-establishing existing connections because if a connection is down then we are already in an exceptional situation
+ * and it doesn't matter much if we stay in this situation a little longer.
+ * <p>
+ * This component does not block on disconnections at all, because a disconnection might need to wait for an ongoing (background) connection
+ * attempt to complete first.
  */
 public class NodeConnectionsService extends AbstractLifecycleComponent {
     private static final Logger logger = LogManager.getLogger(NodeConnectionsService.class);
 
     public static final Setting<TimeValue> CLUSTER_NODE_RECONNECT_INTERVAL_SETTING =
-            positiveTimeSetting("cluster.nodes.reconnect_interval", TimeValue.timeValueSeconds(10), Property.NodeScope);
+        positiveTimeSetting("cluster.nodes.reconnect_interval", TimeValue.timeValueSeconds(10), Property.NodeScope);
+
     private final ThreadPool threadPool;
     private final TransportService transportService;
 
-    // map between current node and the number of failed connection attempts. 0 means successfully connected.
-    // if a node doesn't appear in this list it shouldn't be monitored
-    private ConcurrentMap<DiscoveryNode, Integer> nodes = ConcurrentCollections.newConcurrentMap();
+    // protects changes to targetsByNode and its values but crucially is not held while connecting or disconnecting
+    private final Object mutex = new Object();
```

Comment:
The comment does not fully describe the purpose of this mutex - mutes is also used in ConnectionTarget. Can you add the description of how it's used in ConnectionTarget to the comment?

Reply:
Sure, I expanded "and its value" to clarify in 23cc56d7d6f2fa75fc15115cf04e25159e879806.